bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitHeaterCoolerImpl.java

```diff
@@ -0,0 +1,197 @@
+import io.github.hapjava.characteristics.impl.heatercooler.CurrentHeaterCoolerStateEnum;
+import io.github.hapjava.characteristics.impl.heatercooler.TargetHeaterCoolerStateEnum;
+import io.github.hapjava.characteristics.impl.thermostat.TemperatureDisplayUnitCharacteristic;
+import io.github.hapjava.characteristics.impl.thermostat.TemperatureDisplayUnitEnum;
+import io.github.hapjava.services.impl.HeaterCoolerService;
+
+/**
+ * Implements Heater Cooler
+ *
+ * @author Eugen Freiter - Initial contribution
+ */
+
+public class HomekitHeaterCoolerImpl extends AbstractHomekitAccessoryImpl implements HeaterCoolerAccessory {
+    private final Logger logger = LoggerFactory.getLogger(HomekitHeaterCoolerImpl.class);
+    private final BooleanItemReader activeReader;
+    private final Map<CurrentHeaterCoolerStateEnum, String> currentStateMapping = new EnumMap(
+            CurrentHeaterCoolerStateEnum.class) {
+        {
+            put(CurrentHeaterCoolerStateEnum.INACTIVE, "INACTIVE");
+            put(CurrentHeaterCoolerStateEnum.IDLE, "IDLE");
+            put(CurrentHeaterCoolerStateEnum.HEATING, "HEATING");
+            put(CurrentHeaterCoolerStateEnum.COOLING, "COOLING");
+
+        }
+    };
+    private final Map<TargetHeaterCoolerStateEnum, String> targetStateMapping = new EnumMap(
+            TargetHeaterCoolerStateEnum.class) {
+        {
+            put(TargetHeaterCoolerStateEnum.AUTO, "AUTO");
+            put(TargetHeaterCoolerStateEnum.HEAT, "HEAT");
+            put(TargetHeaterCoolerStateEnum.COOL, "COOL");
+        }
+    };
+
+    public HomekitHeaterCoolerImpl(HomekitTaggedItem taggedItem, List<HomekitTaggedItem> mandatoryCharacteristics,
+            HomekitAccessoryUpdater updater, HomekitSettings settings) throws IncompleteAccessoryException {
+        super(taggedItem, mandatoryCharacteristics, updater, settings);
+        activeReader = new BooleanItemReader(getItem(HomekitCharacteristicType.ACTIVE_STATUS, GenericItem.class),
+                OnOffType.ON, OpenClosedType.OPEN);
+        updateMapping(CURRENT_HEATER_COOLER_STATE, currentStateMapping);
+        updateMapping(TARGET_HEATER_COOLER_STATE, targetStateMapping);
+        final HeaterCoolerService service = new HeaterCoolerService(this);
+        service.addOptionalCharacteristic(new TemperatureDisplayUnitCharacteristic(() -> getTemperatureDisplayUnit(),
+                (value) -> setTemperatureDisplayUnit(value), (callback) -> subscribeTemperatureDisplayUnit(callback),
+                () -> unsubscribeTemperatureDisplayUnit()));
+        getServices().add(service);
+    }
+
+    @Override
+    public CompletableFuture<Double> getCurrentTemperature() {
+        @Nullable
+        final DecimalType state = getStateAs(HomekitCharacteristicType.CURRENT_TEMPERATURE, DecimalType.class);
+        return CompletableFuture.completedFuture(state != null ? convertToCelsius(state.doubleValue()) : 0.0);
```

Comment:
If the underlying state could be a QuantityType then you should use the QuantityType to handle your unit conversions.
```suggestion
        final DecimalType state = getStateAs(HomekitCharacteristicType.CURRENT_TEMPERATURE, DecimalType.class);
        return CompletableFuture.completedFuture(state != null ? convertToCelsius(state.doubleValue()) : 0.0);
```

Reply:
to me it looks like "covertToCelsius" makes already use of QuantityType (AbstractHomekitAccessoryImpl.convertAndRound)

Reply:
I'm just concerned that `getStateAs` would end up converting a QuantityType into a DecimalType thus losing access to the units field.
Is it guaranteed that all of the items you are tracking are using the same units determined in getSettings()? What if the settings change?

Reply:
currently we are using temperature (and conversion from C to F) in thermostat, temperature sensor and with this PR also in HeaterCooler. they all are using the same method.

you are right, the QuantityType gets lost and also settings change will not get propagated to the items and to the home app. the other direction will also not work, i.e. change of units in home app will have no impact on OH settings. So, basically the current code does not support changes of temperature units at runtime. 

it needs to fix. but i would suggest to make a separate enhancement & PR for this and fix it for all 3 accessories with temperature units.

Reply:
I agree, it should be done in another PR.